keras/engine/training.py

```diff
@@ -3680,7 +3680,12 @@ def compile_from_config(self, config):
             return
         config = saving_lib.deserialize_keras_object(config)
         self.compile(**config)
-        if hasattr(self, "optimizer") and self.built:
+        if (
+            hasattr(self, "optimizer")
+            # Exempt legacy optimizers.
+            and isinstance(self.optimizer, optimizer.Optimizer)
```

Comment:
Is it every possible to pass an optimizer that isn't an `Optimizer` instance? Wouldn't that blow up?

Reply:
All custom optimizers that need to be built on deserialization must subclass from `Optimizer` afaik, and legacy optimizers (which do not) are exempted here because they do not have a corresponding `build` method.

Is there any edge case in particular you are referring to? The corresponding CL did not cause any TGP issues.

Reply:
Legacy optimizers simply don't exist in Keras 3, do they?

Reply:
This PR is actually for the Keras 2.15 release, so legacy optimizers still exist in this revision actually (merging into the cherrypick branch).

Reply:
Got it, I had not noticed this was in the 2.15 branch.